src/Features/Core/Portable/AddImport/AbstractAddImportFeatureService.cs

```diff
@@ -376,10 +379,22 @@ private static bool IsHostOrRemoteWorkspace(Project project)
         /// </summary>
         private static bool IsInPackagesDirectory(PortableExecutableReference reference)
         {
-            return ContainsPathComponent(reference, "packages")
-                || ContainsPathComponent(reference, "packs")
-                || ContainsPathComponent(reference, "NuGetFallbackFolder")
-                || ContainsPathComponent(reference, "NuGetPackages");
+            var metadataId = reference.GetMetadataId();
+            if (!s_isInPackagesDirectory.TryGetValue(metadataId, out var result))
+            {
+                result = ComputeIsInPackagesDirectory();
+                s_isInPackagesDirectory.TryAdd(metadataId, result);
+            }
+
+            return result;
+
+            bool ComputeIsInPackagesDirectory()
+            {
+                return ContainsPathComponent(reference, "packages")
```

Comment:
I considered that, but it would still be bad.  we'd still then do all the string operations (even if not allocating) over and over again, even though the results never change.  Consider if we have 100 references, each with long paths that have to be searched for these components.  Why do that over and over again when nothing changes and it can be O(1) lookup in the future?